deepcell/applications/application.py

```diff
@@ -0,0 +1,164 @@
+                              postprocess_kwargs={},
+                              debug=False):
+        """Generates a labeled image of the input running prediction with
+        appropriate pre and post processing functions.
+
+        Input images are required to have 4 dimensions [batch, x, y, channel]. Additional
+        empty dimensions can be added using `np.expand_dims`
+
+        Args:
+            image (np.array): Input image with shape [batch, x, y, channel]
+            batch_size (int, optional): Number of images to predict on per batch. Defaults to 4.
+            image_mpp (float, optional): Microns per pixel for the input image. Defaults to None.
+            preprocess_kwargs (dict, optional): Kwargs to pass to preprocessing function.
+                Defaults to {}.
+            postprocess_kwargs (dict, optional): Kwargs to pass to postprocessing function.
+                Defaults to {}.
+            debug (bool, optional): If True, returns intermediate outputs of data processing.
+                Defaults to False.
+
+        Raises:
+            ValueError: Input data must have 4 dimensions, [batch, x, y, channel
+
+        Returns:
+            np.array: Labeled image, if debug is False.
+                If debug is True, returns (image, tiles, output_tiles, output_images, label_image)
+        """
+
+        # Check input size of image
+        if len(image.shape) != self.required_rank:
+            raise ValueError('Input data must have {} dimensions'
+                             'Input data only has {} dimensions'.format(str(self.required_rank),
+                                                                        str(len(image.shape))))
+
+        # Resize image if necessary
+        if (image_mpp is not None) & (image_mpp != self.model_mpp):
+            original_shape = image.shape
+            scale_factor = image_mpp / self.model_mpp
+            new_shape = (int(image.shape[1] / scale_factor),
+                         int(image.shape[2] / scale_factor))
+            image = resize(image, new_shape, data_format='channels_last')
+        else:
+            original_shape = None
+
+        # Preprocess image
+        if self.preprocessing_fn is not None:
+            image = self.preprocessing_fn(image, **preprocess_kwargs)
+
+        # Tile images, needs 4d
+        tiles, tiles_info = tile_image(image, model_input_shape=self.model_image_shape)
+
+        # Run images through model
+        output_tiles = self.model.predict(tiles, batch_size=batch_size)
+
+        # Untile images
+        output_images = [untile_image(o, tiles_info, model_input_shape=self.model_image_shape,
+                                      dtype=o.dtype) for o in output_tiles]
+
+        # Postprocess predictions to create label image
+        if self.postprocessing_fn is not None:
+            label_image = self.postprocessing_fn(output_images, **postprocess_kwargs)
+        else:
+            label_image = output_images[0]
```

Comment:
why only the first value?

Reply:
When we untile we use list comprehension, so the output always ends up in a list. If there is only a single image as output then line 155 essentially takes it out of the list. 

The inherent assumption is that after running postprocessing there should only be a single image. We rely on that assumption during the resize test. I'm not sure if it's appropriate to generalize this further. Thoughts?

Reply:
This is no longer an issue after creating data processing utility methods.